divintseg/_core.py

```diff
@@ -422,3 +422,144 @@ def similarity(
+    0  Region 1  0.83333
+    1  Region 2      0.1
+
+    Let's look at what this example computed. First, we have
+    to see how likely each person in group S is to see other
+    members of their own group in their subregion. This is as
+    follows:
+
+    +----------+-------------+-----------------------+
+    | Region   | Subregion   | Likelihood of an S    |
+    +==========+=============+=======================+
+    | Region 1 | Subregion A | 100 / (100 + 0) = 1.0 |
+    +----------+-------------+-----------------------+
+    | Region 1 | Subregion B | 50 / (50 + 50) = 0.5  |
+    +----------+-------------+-----------------------+
+    | Region 2 | Subregion C | 0 / (0 + 100) = 0.0   |
+    +----------+-------------+-----------------------+
+    | Region 2 | Subregion D | 0 / (0 + 50) = 0.0    |
+    +----------+-------------+-----------------------+
+    | Region 2 | Subregion E | 10 / (10 + 90) = 0.1  |
+    +----------+-------------+-----------------------+
+
+    Next, we can compute the fraction of all S's in
+    each subregion of each region. There are 150 S's
+    in Region 1 and 10 S's in region 2, therefore, we have:
+
+    +----------+-------------+--------------------------------+
+    | Region   | Subregion   | Fraction of all As in Region   |
+    +==========+=============+================================+
+    | Region 1 | Subregion A | 100 / 150 = 0.6667             |
+    +----------+-------------+--------------------------------+
+    | Region 1 | Subregion B | 50 / 150 = 0.3333              |
+    +----------+-------------+--------------------------------+
+    | Region 2 | Subregion C | 0 / 10 = 0.0000                |
+    +----------+-------------+--------------------------------+
+    | Region 2 | Subregion D | 0 / 10 = 0.0000                |
+    +----------+-------------+--------------------------------+
+    | Region 2 | Subregion E | 10 / 10 = 1.0000               |
+    +----------+-------------+--------------------------------+
+
+    Finally, for each subregion, we multiply these together and
+    add them up the values for the subregions in each region.
+    For Region 1, we get
+
+    .. math::
+        (0.6667 * 1.0) + (0.3333 * 0.5) = 0.8333.
+
+    For Region 2, we get
+
+    .. math::
+        0.0 * 0.0 + 0.0 * 0.0 + 1.000 * 0.1 = 0.1.
+
+    Note that the implentation may not do this math exactly as
+    specified here, but it will do something equivalent.
+    """
+
+    df_grouped = df_communities.groupby([by, over], as_index=False).sum(
+        numeric_only=True
+    )
+    likelihood = df_grouped[group_name] / df_grouped.sum(
+        axis="columns", numeric_only=True
+    )
+
+    def get_region_pop(row):
+        # Used with apply to add region population columns
+        row["Region Population"] = df_grouped[df_grouped[by] == row[by]].sum(
+            numeric_only=True
+        )[group_name]
+        return row
+
+    df_grouped = df_grouped.apply(get_region_pop, axis=1)
```

Comment:
As you may have already learned in the time you have spent with Pandas, there are often many different ways to compute the same thing. Some are more or less concise than others; some are more or less efficient than others. So it's often important to think a little about what choice you make and why. And with experience, the choices become easier and more familiar.

In this case, let's look as what you are doing here to get the total population of members of `group_name` in each region. When you do the `apply`, you are calling `get_region_pop` once for each row. And inside that, you are finding all rows that have the the same `by` as the row and summing them up. This means that you are doing the same sum multiple times, once for each row in the rows being summed, because you are calling `get_region_pop` for each row. Beyond that, you are doing the sum for all numeric columns, then selecting `[group_name]` and this throwing away the sums you just computed for all the other columns.

What if, instead, you did something like
```
    region_group_population = df_grouped.groupby(by)[group_name].sum(numeric_only=True)
```

What we are doing here is only doing the sum once for each value of the `by` column and only doing it for `group_name`, because we select with `[group_name]` before the sum.

Notice that i have a series now, not a new column in `df_grouped`. So it is a little different. I don't want to give away the whole plot, so I will leave it as an exercise to you how to merge/join this back to `df_grouped` and do the rest of the computation.

I'm happy to help further if you have more questions.

Reply:
Thanks again for the quick and thorough feedback, along with the Github tip; I haven't had much experience with pull requests so that was really helpful! I have just committed my code that calculated region population more efficiently in the way that you mentioned. Please let me know if there is anything else you would like me to fix or add. Thank you so much for your time!